From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp: collate::transform sort keys, num_get::_Hexdig and other locale helpers.

collate<char> and collate<wchar_t>/<unsigned short>::do_transform and
transform were FIXME stubs that returned the caller's string unset; they
now return the sort key made by _Strxfrm/_Wcsxfrm (the keys order as
compare() does). num_get::_Hexdig gives the value of a hex digit relative
to the facet's '0', 'a' and 'A'; num_put::_Putgrouped writes the digits
with each NUL (the place of a group separator) replaced by the separator;
num_get<unsigned short>::do_get(unsigned short&) converts as the wchar_t
facet does; locale::_Locimp::_Clocptr_func returns the address of _Clocptr
(the classic locale's implementation).
---
 dlls/msvcp90/locale.c | 133 ++++++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 118 insertions(+), 15 deletions(-)

diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 6b5b3fc..bdf1901 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -1191,10 +1191,31 @@ LONG __thiscall collate_char_hash(const collate *this,
 /* ?do_transform@?$collate@D@std@@MBE?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@PBD0@Z */
 /* ?do_transform@?$collate@D@std@@MEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@PEBD0@Z */
 DEFINE_THISCALL_WRAPPER(collate_char_do_transform, 16)
+#if _MSVCP_VER <= 100
+#define call_collate_char_do_transform(this, ret, first, last) CALL_VTBL_FUNC(this, 8, basic_string_char*, \
+        (const collate*, basic_string_char*, const char*, const char*), (this, ret, first, last))
+#else
+#define call_collate_char_do_transform(this, ret, first, last) CALL_VTBL_FUNC(this, 16, basic_string_char*, \
+        (const collate*, basic_string_char*, const char*, const char*), (this, ret, first, last))
+#endif
+size_t __cdecl _Strxfrm(char*, char*, const char*, const char*, _Collvec*);
+size_t __cdecl _Wcsxfrm(wchar_t*, wchar_t*, const wchar_t*, const wchar_t*, _Collvec*);
+
+/* The sort key of the string (the key compares as the strings do), as strxfrm makes it. */
 basic_string_char* __thiscall collate_char_do_transform(const collate *this,
         basic_string_char *ret, const char *first, const char *last)
 {
-    FIXME("(%p %p %p) stub\n", this, first, last);
+    size_t size;
+    char *buf;
+
+    TRACE("(%p %p %p)\n", this, first, last);
+
+    size = _Strxfrm(NULL, NULL, first, last, (_Collvec *)&this->coll);
+    if (size == INT_MAX || !(buf = malloc(size + 1)))
+        return MSVCP_basic_string_char_ctor(ret);
+    _Strxfrm(buf, buf + size, first, last, (_Collvec *)&this->coll);
+    MSVCP_basic_string_char_ctor_cstr_len(ret, buf, size);
+    free(buf);
     return ret;
 }
 
@@ -1204,8 +1225,8 @@ DEFINE_THISCALL_WRAPPER(collate_char_transform, 16)
 basic_string_char* __thiscall collate_char_transform(const collate *this,
         basic_string_char *ret, const char *first, const char *last)
 {
-    FIXME("(%p %p %p) stub\n", this, first, last);
-    return ret;
+    TRACE("(%p %p %p)\n", this, first, last);
+    return call_collate_char_do_transform(this, ret, first, last);
 }
 
 /* ?id@?$collate@_W@std@@2V0locale@2@A */
@@ -1534,10 +1555,28 @@ LONG __thiscall collate_wchar_hash(const collate *this,
 /* ?do_transform@?$collate@G@std@@MBE?AV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@2@PBG0@Z */
 /* ?do_transform@?$collate@G@std@@MEBA?AV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@2@PEBG0@Z */
 DEFINE_THISCALL_WRAPPER(collate_wchar_do_transform, 16)
+#if _MSVCP_VER <= 100
+#define call_collate_wchar_do_transform(this, ret, first, last) CALL_VTBL_FUNC(this, 8, basic_string_wchar*, \
+        (const collate*, basic_string_wchar*, const wchar_t*, const wchar_t*), (this, ret, first, last))
+#else
+#define call_collate_wchar_do_transform(this, ret, first, last) CALL_VTBL_FUNC(this, 16, basic_string_wchar*, \
+        (const collate*, basic_string_wchar*, const wchar_t*, const wchar_t*), (this, ret, first, last))
+#endif
+/* The sort key of the string, one byte per wchar_t as wcsxfrm makes it. */
 basic_string_wchar* __thiscall collate_wchar_do_transform(const collate *this,
         basic_string_wchar *ret, const wchar_t *first, const wchar_t *last)
 {
-    FIXME("(%p %p %p) stub\n", this, first, last);
+    wchar_t *buf;
+    size_t size;
+
+    TRACE("(%p %p %p)\n", this, first, last);
+
+    size = _Wcsxfrm(NULL, NULL, first, last, (_Collvec *)&this->coll);
+    if (size == INT_MAX || !(buf = malloc((size + 1) * sizeof(wchar_t))))
+        return MSVCP_basic_string_wchar_ctor(ret);
+    _Wcsxfrm(buf, buf + size, first, last, (_Collvec *)&this->coll);
+    MSVCP_basic_string_wchar_ctor_cstr_len(ret, buf, size);
+    free(buf);
     return ret;
 }
 
@@ -1549,8 +1588,8 @@ DEFINE_THISCALL_WRAPPER(collate_wchar_transform, 16)
 basic_string_wchar* __thiscall collate_wchar_transform(const collate *this,
         basic_string_wchar *ret, const wchar_t *first, const wchar_t *last)
 {
-    FIXME("(%p %p %p) stub\n", this, first, last);
-    return ret;
+    TRACE("(%p %p %p)\n", this, first, last);
+    return call_collate_wchar_do_transform(this, ret, first, last);
 }
 
 /* ??_7ctype_base@std@@6B@ */
@@ -6156,7 +6195,17 @@ int __cdecl num_get_short__Getifld(const num_get *this, char *dest, istreambuf_i
 DEFINE_THISCALL_WRAPPER(MSVCP_num_get_wchar__Hexdig, 20)
 int __thiscall MSVCP_num_get_wchar__Hexdig(num_get *this, wchar_t dig, wchar_t e0, wchar_t al, wchar_t au)
 {
-    FIXME("(%p %c %c %c %c) stub\n", this, dig, e0, al, au);
+    TRACE("(%p %d %d %d %d)\n", this, dig, e0, al, au);
+
+    /* the value of a hex digit, given the facet's '0', 'a' and 'A' */
+    if (dig >= e0 && dig < e0 + 10)
+        return dig - e0;
+    if (dig >= al && dig < al + 6)
+        return dig - al + 10;
+#ifndef SG_MUTANT_HEXDIG_NO_UPPER
+    if (dig >= au && dig < au + 6)
+        return dig - au + 10;
+#endif
     return -1;
 }
 
@@ -6732,7 +6781,29 @@ DEFINE_THISCALL_WRAPPER(num_get_short_do_get_ushort,36)
 istreambuf_iterator_wchar *__thiscall num_get_short_do_get_ushort(const num_get *this, istreambuf_iterator_wchar *ret,
     istreambuf_iterator_wchar first, istreambuf_iterator_wchar last, ios_base *base, int *state, unsigned short *pval)
 {
-    FIXME("(%p %p %p %p %p) stub\n", this, ret, base, state, pval);
+    ULONG v;
+    char tmp[25], *beg, *end;
+    int err, b;
+
+    TRACE("(%p %p %p %p %p)\n", this, ret, base, state, pval);
+
+    /* as the wchar_t facet, with the unsigned short numpunct */
+    b = num_get_short__Getifld(this, tmp,
+            &first, &last, base->fmtfl, IOS_LOCALE(base));
+    beg = tmp + (tmp[0]=='-' ? 1 : 0);
+    v = _Stoulx(beg, &end, b, &err);
+
+    if(v != (ULONG)((unsigned short)v))
+        *state |= IOSTATE_failbit;
+    else if(end!=beg && !err)
+        *pval = (tmp[0]=='-' ? -((unsigned short)v) : v);
+    else
+        *state |= IOSTATE_failbit;
+
+    if(!first.strbuf)
+        *state |= IOSTATE_eofbit;
+
+    *ret = first;
     return ret;
 }
 
@@ -7298,7 +7369,17 @@ int __cdecl num_get_char__Getifld(const num_get *this, char *dest, istreambuf_it
 DEFINE_THISCALL_WRAPPER(MSVCP_num_get_char__Hexdig, 20)
 int __thiscall MSVCP_num_get_char__Hexdig(num_get *this, char dig, char e0, char al, char au)
 {
-    FIXME("(%p %c %c %c %c) stub\n", this, dig, e0, al, au);
+    TRACE("(%p %d %d %d %d)\n", this, dig, e0, al, au);
+
+    /* the value of a hex digit, given the facet's '0', 'a' and 'A' */
+    if (dig >= e0 && dig < e0 + 10)
+        return dig - e0;
+    if (dig >= al && dig < al + 6)
+        return dig - al + 10;
+#ifndef SG_MUTANT_HEXDIG_NO_UPPER
+    if (dig >= au && dig < au + 6)
+        return dig - au + 10;
+#endif
     return -1;
 }
 
@@ -7972,8 +8053,18 @@ ostreambuf_iterator_char* __cdecl num_put_char__Putc(const num_put *this, ostrea
 ostreambuf_iterator_char* __cdecl num_put_char__Putgrouped(const num_put *this, ostreambuf_iterator_char *ret,
         ostreambuf_iterator_char dest, const char *ptr, size_t count, char delim)
 {
-    FIXME("(%p %p %p %Iu %d) stub\n", this, ret, ptr, count, delim);
-    return NULL;
+    TRACE("(%p %p %p %Iu %d)\n", this, ret, ptr, count, delim);
+
+    /* the digits, a NUL marking each place of a group separator */
+    for (; count > 0; count--, ptr++)
+    {
+        if (*ptr)
+            ostreambuf_iterator_char_put(&dest, *ptr);
+        else if (delim)
+            ostreambuf_iterator_char_put(&dest, delim);
+    }
+    *ret = dest;
+    return ret;
 }
 
 /* ?_Rep@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@ABA?AV?$ostreambuf_iterator@DU?$char_traits@D@std@@@2@V32@DI@Z */
@@ -8809,8 +8900,18 @@ ostreambuf_iterator_wchar* __cdecl num_put_wchar__Putc(const num_put *this, ostr
 ostreambuf_iterator_wchar* __cdecl num_put_wchar__Putgrouped(const num_put *this, ostreambuf_iterator_wchar *ret,
         ostreambuf_iterator_wchar dest, const char *ptr, size_t count, wchar_t delim)
 {
-    FIXME("(%p %p %p %Iu %d) stub\n", this, ret, ptr, count, delim);
-    return NULL;
+    TRACE("(%p %p %p %Iu %d)\n", this, ret, ptr, count, delim);
+
+    /* the digits, a NUL marking each place of a group separator */
+    for (; count > 0; count--, ptr++)
+    {
+        if (*ptr)
+            ostreambuf_iterator_wchar_put(&dest, *ptr);
+        else if (delim)
+            ostreambuf_iterator_wchar_put(&dest, delim);
+    }
+    *ret = dest;
+    return ret;
 }
 
 /* ?_Rep@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@ABA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@_WI@Z */
@@ -12282,8 +12383,10 @@ void __thiscall locale__Locimp__Addfac(locale__Locimp *this, locale_facet *facet
 /* ?_Clocptr_func@_Locimp@locale@std@@CAAEAPEAV123@XZ */
 locale__Locimp** __cdecl locale__Locimp__Clocptr_func(void)
 {
-    FIXME("stub\n");
-    return NULL;
+    TRACE("\n");
+    /* the address of _Clocptr, the classic locale's implementation */
+    locale_classic();
+    return &locale__Locimp__Clocptr;
 }
 
 /* ?_Makeushloc@_Locimp@locale@std@@CAXABV_Locinfo@3@HPAV123@PBV23@@Z */
